CMake: fix targets linker script/drivers inclusion

0xc0170 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Thanks @ladislas for highlighting this. I refactored slightly the targets code, a driver should have one CMakelists (if updating, only one file needs an update). 

Adding linker/startup does not need any function, just add the files within the same ifdef toolchain. I removed unused cmakelists.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [x] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

0xc0170 committed:
CMake k64f: align linker/startup

0xc0170 committed:
CMake k64f: refactor device files

Make consistent k64f cmakelists (a preparation for a library soon).

0xc0170 committed:
CMake stm32f439: align linker/startup

0xc0170 committed:
CMake stm32f4: fix cube driver inclusion

0xc0170 committed:
CMake stm32f4: add cube driver on the top

0xc0170 committed:
CMake stm32f401: align linker/startup

0xc0170 committed:
CMake k66f: align linker/startup

0xc0170 committed:
CMake stm32f3: align linker/startup

0xc0170 committed:
CMake nrf52480: align linker/startup

0xc0170 committed:
CMake stm32l4: fix cube driver inclusion

ladislas commented on targets/TARGET_STM/TARGET_STM32F4/STM32Cube_FW/CMakeLists.txt:
```diff
@@ -1,15 +1,107 @@
 
 target_include_directories(mbed-core
     INTERFACE
         ${CMAKE_CURRENT_SOURCE_DIR}
         ${CMAKE_CURRENT_SOURCE_DIR}/CMSIS
+        ${CMAKE_CURRENT_SOURCE_DIR}/STM32F4xx_HAL_Driver
+        ${CMAKE_CURRENT_SOURCE_DIR}/STM32F4xx_HAL_Driver/Legacy
```
This is the same as here:

https://github.com/ARMmbed/mbed-os/pull/13835/files#diff-7c98b07952b498f1b0449bd8fdc3c0f1c2dac26ff61884d00a56c8aa8af36e41R26-R29

I think it can be removed from the TARGET_STM32F4/CMakeLists.txt

0xc0170 replied:
Is this referring to `CMAKE_CURRENT_SOURCE_DIR` or something else? the diff points to k66f file

ladislas replied:
Yeah sorry, the copy/paste in diff is not working well when you expand the files.

I meant here:

https://github.com/0xc0170/mbed-os/blob/24301485ea6f350ade61eb8f4486f5b2dd7bc0e8/targets/TARGET_STM/TARGET_STM32F4/CMakeLists.txt#L26-L29

```cmake
target_include_directories(mbed-core
    INTERFACE
        ${CMAKE_CURRENT_SOURCE_DIR}
        ${CMAKE_CURRENT_SOURCE_DIR}/STM32Cube_FW
        ${CMAKE_CURRENT_SOURCE_DIR}/STM32Cube_FW/CMSIS
        ${CMAKE_CURRENT_SOURCE_DIR}/STM32Cube_FW/STM32F4xx_HAL_Driver
        ${CMAKE_CURRENT_SOURCE_DIR}/STM32Cube_FW/STM32F4xx_HAL_Driver/Legacy
)
```

is the same as here:

https://github.com/0xc0170/mbed-os/blob/24301485ea6f350ade61eb8f4486f5b2dd7bc0e8/targets/TARGET_STM/TARGET_STM32F4/STM32Cube_FW/CMakeLists.txt#L103-L106

```cmake
target_include_directories(mbed-core
    INTERFACE
        ${CMAKE_CURRENT_SOURCE_DIR}
        ${CMAKE_CURRENT_SOURCE_DIR}/CMSIS
        ${CMAKE_CURRENT_SOURCE_DIR}/STM32F4xx_HAL_Driver
        ${CMAKE_CURRENT_SOURCE_DIR}/STM32F4xx_HAL_Driver/Legacy
)
```

in terms of includes.

0xc0170 replied:
+1, fixed now

ladislas commented on targets/TARGET_STM/TARGET_STM32L4/STM32Cube_FW/CMakeLists.txt:
```diff
@@ -1,15 +1,120 @@
-        ${CMAKE_CURRENT_SOURCE_DIR}
-	CMSIS
+        .
+        CMSIS
+        STM32L4xx_HAL_Driver
+        STM32L4xx_HAL_Driver/Legacy
```
F4 uses `${CMAKE_CURRENT_SOURCE_DIR}/` -- I think we should be consistent.

I personally prefer without the `${CMAKE_CURRENT_SOURCE_DIR}`

0xc0170 replied:
We will refactor entire codebase (I only changed what I touched here)

ladislas replied:
👍

ladislas reviewed (commented):
Thanks for addressing my questions @0xc0170  :)

0xc0170 committed:
CMake stm32f4: remove cube include fix

hugueskamba commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_K66F/CMakeLists.txt:
```diff
@@ -5,12 +5,24 @@ if("FRDM" IN_LIST MBED_TARGET_LABELS)
+    set(STARTUP_FILE device/TOOLCHAIN_IAR/startup_MK66F18.S)
+    set(LINKER_FILE device/TOOLCHAIN_IAR/MK66FN2M0xxx18.icf)
+endif()
+target_sources(mbed-core INTERFACE ${STARTUP_FILE})
```
```suggestion
```

0xc0170 replied:
Fixed

hugueskamba commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_K66F/CMakeLists.txt:
```diff
@@ -21,6 +33,8 @@ target_sources(mbed-core
         trng_api.c
```
```suggestion
        ${STARTUP_FILE}
        trng_api.c
```

0xc0170 replied:
Fixed

hugueskamba commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52840/CMakeLists.txt:
```diff
@@ -5,16 +5,31 @@ if("NRF52840_DK" IN_LIST MBED_TARGET_LABELS)
 
 target_sources(mbed-core
     INTERFACE
         PeripheralPins.c
```
```suggestion
        ${STARTUP_FILE}
        PeripheralPins.c
```

0xc0170 replied:
Fixed

hugueskamba commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52840/CMakeLists.txt:
```diff
@@ -5,16 +5,31 @@ if("NRF52840_DK" IN_LIST MBED_TARGET_LABELS)
+    set(LINKER_FILE device/TOOLCHAIN_IAR/nRF52840.icf)
+endif()
+
+target_sources(mbed-core INTERFACE ${STARTUP_FILE})
```
```suggestion
```

0xc0170 replied:
Fixed

hugueskamba commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52840/CMakeLists.txt:
```diff
@@ -5,16 +5,31 @@ if("NRF52840_DK" IN_LIST MBED_TARGET_LABELS)
-)
 
-add_subdirectory(device)
+        device/system_nrf52840.c
```
```suggestion
        device/cmsis_nvic.c
        device/system_nrf52840.c
```

0xc0170 replied:
fixed

hugueskamba reviewed (dismissed):
Thank you for the changes.
One file was missed, see `targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52840/CMakeLists.txt`

0xc0170 committed:
CMake targets: fix how we include startup

Styling fix. Plus also one target missed in the previous run.

0xc0170 committed:
CMake targets: use current source dir for the linker script

As the linker scripts is being invoked from the main Mbed OS, we need the full path to the file.
